### ISSUE TITLE
Support for MSBuildGlob serialization round-tripping

### ISSUE THREAD

**Issue by AArnott** [dotnet/msbuild#6926]
Given an MSBuildGlob (and presumably its variant classes), I need to be able to serialize and deserialize it. I cannot today because a given instance does not expose properties that match any constructor.

The scenario is RPC: I would like to send MSBuildGlob objects as arguments over RPC for VS caching.

This could be mitigated by simply adding a property whose value is the originally parsed string, since if I can acquire that, the remote process can simply pass it to Parse again.

Note that some globs are composite (e.g. `*.cs;*.vb;!notme.cs`) so however that is represented in msbuild glob classes needs to be reconstructible over RPC as well.

**Comment by BenVillalobos** [comment]
@AArnott I think [this](https://github.com/dotnet/msbuild/pull/6694#issuecomment-931498821) was what he was referring to?

**Comment by AArnott** [comment]
Copying it here, I guess you mean this comment:

> The latest proposal is to expose the expanded glob as a list of strings that can be serialized or used to deserialize back to MSBuildGlob.

I *do* want the expanded glob, but that's tracked by #6927.

The second part of the above comment ("deserialize back to MSBuildGlob") doesn't make any sense to me as one cannot deserialized a list of file paths into an MSBuildGlob object. I don't want or need that. I want to take an MSBuildGlob object, serialize its content (the patterns -- not the expanded product), then deserialize that data back into an `MSBuildGlob` object.

**Comment by Forgind** [comment]
@arkalyanms

**Comment by arkalyanms** [comment]
The list of strings in the comment refers to the parts of the glob for deduping not the list of the files. 

>> I want to take an MSBuildGlob object, serialize its content (the patterns -- not the expanded product), then deserialize that data back into an MSBuildGlob object.

That sounds the same as #6694, but if you feel that draft is not solve the problem, we can abandon that PR (since CPS has a reconstructions from filespec that unblocks it) and use this as the primary to drive your requirement.

**Comment by AArnott** [comment]
@arkalyanms, No, #6694 does not satisfy the requirement tracked by this issue. That merely exposes a regex, but a regex cannot be used to recreate an MSBuildGlob, and this issue is about serializing an MSBuildGlob such that it can be deserialized.

**Comment by arkalyanms** [comment]
Then #6694 is obsolete at this point. We can replace that with this issue. @Forgind fyi.

@AArnott Perhaps this has been answered. But remind me, what do you intend to do with the glob post deserialization? 

Perhaps your use case falls outside the IMSBuildGlob contract? #6694 was only meant to cover this. The glob gets treated as a black box that answers matching questions around whether a file is covered by the glob, which is the project system scenario.

**Comment by AArnott** [comment]
@arkalyanms  Once I can deserialize the glob (this issue), I intend to use it for #6927.
